File: unDirectedGraph.js
class Node {
  constructor(nodeName) {
    this.nodeName = nodeName;
  }
}

class Edge {
  constructor(fromNode, toNode, weight) {
    this.fromNode = fromNode;
    this.toNode = toNode;
    this.weight = weight;
  }
}

class Graph {
  constructor() {
    this.nodesHashMap = {};
    this.adjacencyHashMap = {};
  }

  isNodePresent = (nodeName) => {
    return this.nodesHashMap[nodeName];
  };

  isNodePresentInAdjacencyList = (nodeName) => {
    return this.adjacencyHashMap[nodeName];
  };

  addNode = (nodeName) => {
    if (this.isNodePresent(nodeName)) return;
    else {
      this.nodesHashMap[nodeName] = new Node(nodeName); // {"A" : NodeA}
      let adjacencyList = [];
      this.adjacencyHashMap[nodeName] = adjacencyList; // {NodeA : []}
    }
  };

  addEdge = (fromNode, toNode, weight) => {
    if (!(this.isNodePresent(fromNode) && this.isNodePresent(toNode))) return;

    this.adjacencyHashMap[fromNode].push(
      new Edge(this.nodesHashMap[fromNode], this.nodesHashMap[toNode], weight)
    ); // {NodeA : [EdgeAB, ...]}

    this.adjacencyHashMap[toNode].push(
      new Edge(this.nodesHashMap[toNode], this.nodesHashMap[fromNode], weight)
    ); // {NodeB : [EdgeBA, ...]}
  };

  cycleDetectionPublic = () => {
    let visitedSet = new Set();
    for (let node in this.nodesHashMap) {
      if (
        !visitedSet.has(this.nodesHashMap[node]) &&
        this.cycleDetectionPrivate(this.nodesHashMap[node], null, visitedSet)
      )
        return true;
    }
    return false;
  };

  cycleDetectionPrivate = (node, parent, visitedSet) => {
    visitedSet.add(node);

    for (let neighbour of this.adjacencyHashMap[node.nodeName]) {
      if (neighbour.toNode == parent) continue;
      if (
        visitedSet.has(neighbour.toNode) ||
        this.cycleDetectionPrivate(neighbour.toNode, node, visitedSet)
      )
        return true;
    }
    return false;
  };
}

let udGraph = new Graph();
udGraph.addNode("B");
udGraph.addNode("A");
udGraph.addNode("C");
udGraph.addNode("D");
udGraph.addEdge("A", "B", 2);
udGraph.addEdge("B", "D", 5);
udGraph.addEdge("D", "C", 10);
// udGraph.addEdge("C", "A", 1);

console.log(udGraph.adjacencyHashMap["A"]);
// console.log(udGraph.nodesHashMap);
// console.log(udGraph.cycleDetectionPublic());

File: unDirectedGraph.py
class Node:
    def __init__(self, nodeName):
        self.nodeName =  nodeName
    def __repr__(self):
        reprResult = "Node {{nodeName: {}}}".format(self.nodeName)
        return reprResult

class Edge:
    def __init__(self, fromNode, toNode, weight):
        self.fromNode = fromNode
        self.toNode =  toNode
        self.weight =  weight
    def __repr__(self):
        reprResult = "Edge {{fromNode: {}, toNode: {}, weight: {}}}".format(self.fromNode, self.toNode, self.weight)
        return reprResult
        
class Graph:
    def __init__(self):
        self.nodesHashMap = {}
        self.adjacencyHashMap = {}
    def __repr__(self):
        return "Graph"

    def isNodePresent(self, nodeName):
        return self.nodesHashMap.get(nodeName)
    
    def addNode(self, nodeName):
        if self.isNodePresent(nodeName):
            return
        else:
            self.nodesHashMap.update({nodeName: Node(nodeName)}) # {"A" : NodeA}
            adjacencyList = []
            self.adjacencyHashMap.update({nodeName: adjacencyList}) # {NodeA : []}

    def addEdge(self, fromNode, toNode, weight):
        if not (self.isNodePresent(fromNode) and self.isNodePresent(toNode)):
            return
        
        self.adjacencyHashMap.get(fromNode).append(Edge(self.nodesHashMap[fromNode], self.nodesHashMap[toNode], weight)) # {Node A : [EdgeAB, ...]}
        self.adjacencyHashMap.get(toNode).append(Edge(self.nodesHashMap[toNode], self.nodesHashMap[fromNode], weight)) # {Node B : [EdgeBA, ...]}

    def cycleDetectionPublic(self):
        visitedSet = set()
        for node in self.nodesHashMap:
            if not self.nodesHashMap[node] in visitedSet and self.cycleDetectionPrivate(self.nodesHashMap[node], None, visitedSet):
                return True

        return False

    def cycleDetectionPrivate(self, node, parent, visitedSet):
        visitedSet.add(node)

        for neighbour in self.adjacencyHashMap[node.nodeName]:
            if neighbour.toNode == parent:
                continue
            if neighbour.toNode in visitedSet or (self.cycleDetectionPrivate(neighbour.toNode, node, visitedSet)):
                return True
        
        return False 


udGraph = Graph()
udGraph.addNode("B")
udGraph.addNode("A")
udGraph.addNode("C")
udGraph.addNode("D")
udGraph.addEdge("A", "B", 5)
udGraph.addEdge("B", "D", 1)
udGraph.addEdge("D", "C", 10)
udGraph.addEdge("A", "C", 3)

print(udGraph.cycleDetectionPublic())
# print(udGraph.adjacencyHashMap["A"])
# print(udGraph.nodesHashMap)
# print(udGraph.adjacencyHashMap["A"])